### PR TITLE
App Router: Support user-defined paths for predefined pages

### PR DESCRIPTION
**Pull request by johnnyomair**
This is achieved by using a combination of rewrites and redirects in the middleware:

- Rewrite from the page tree node path (e.g., `/aktuelles`) to the code path (e.g., `/news`)
- Redirect from the code path (e.g., `/news`) to the page tree node path (e.g., `/aktuelles`)
---

https://github.com/user-attachments/assets/08b21a20-648e-4566-9728-7dbb23e6dbc5

### CODE REVIEW

**Comment by johnnyomair** [comment]
Need to fix linking for NewsLinkBlock

**Comment by johnnyomair** [comment]
> Need to fix linking for NewsLinkBlock

Done in #2284 since it isn't concerned to the user-defined-paths

**Review by nsams — APPROVED** [review]
should target main i guess

**Comment by johnnyomair** [comment]
@nsams please review, thanks!

File: demo/site/src/middleware/predefinedPages.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                pages.push({
+                    defaultPath: `/${language}${predefinedPagePaths[node.document.type]}`,
+                    customPath: `/${language}${node.path}`,
+                });
+            }
+        }
+    }
+
+    return memoryCache.wrap(key, async () => {
+        return pages;
```
Comment (nsams) [inline demo/site/src/middleware/predefinedPages.ts:73]: This cache is not working correctly, you are fetching every time and caching only the response variable.

File: demo/site/src/middleware/predefinedPages.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { gql } from "@comet/cms-site";
+import { languages } from "@src/config";
```
Comment (nsams) [inline demo/site/src/middleware/predefinedPages.ts:2]: that have to change when we have siteConfig in demo (we should have asap!!!)

File: demo/site/src/middleware/predefinedPages.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }
+`;
+
+const graphQLFetch = createGraphQLFetch();
+
+async function fetchPredefinedPages(scope: { domain: string }) {
+    const key = `predefinedPages-${JSON.stringify(scope)}`;
+
+    const pages: Array<{ defaultPath: string; customPath: string }> = [];
```
Comment (nsams) [inline demo/site/src/middleware/predefinedPages.ts:55]: naming is not easy to understand imho, maybe:

- defaultPath -> codePath
- customPath -> configuredPath

(or you have a better idea)

File: demo/site/src/middleware/predefinedPages.ts (hunk 1 above)
Comment (johnnyomair) [reply demo/site/src/middleware/predefinedPages.ts:73]: You're right, missed that. Fixed: 52fa2bd7d5321bca402a8d8e3a44808405395d51

File: demo/site/src/middleware/predefinedPages.ts (hunk 3 above)
Comment (johnnyomair) [reply demo/site/src/middleware/predefinedPages.ts]: Changed to codePath and pageTreeNodePath: 2011f235db8dcf397b0f4aa9a581886f636535a6

**Comment by johnnyomair** [comment]
@nsams please review, thanks!

**Review by nsams — APPROVED** [review]

**Pull request merged**